File: evidence/cpm1/before-a/modules/sow/apps/desktop/ipc/envelope.js
"use strict";
/**
 * Node mirror of control_plane/ipc/envelope.py — the shell side of the D-IPC-01 contract.
 *
 * The Python reference is authoritative; this file exists ONLY so the desktop shell
 * (a Node/Chromium main process) computes byte-identical envelopes and integrity tags.
 * Two independent checks must pass before anything acts on a message (fail-closed):
 *   1. structural: it is a well-formed envelope@1.0 (the gateway re-validates against
 *      schemas/message.schema.json — we build to that shape, never trust our own claim);
 *   2. integrity: integrity === hmac_sha256(shared_key, canonical(envelope minus integrity)).
 *
 * U458 — WHY (2) NOW SAYS "envelope" AND NOT "payload". W-41 widened the signed input in the Python
 * reference from `payload` to the whole envelope, so `from_node`, `type`, `task_id`, `msg_id` and
 * `ts` stopped being rewritable in flight. THIS FILE WAS NOT CHANGED WITH IT. Both halves passed
 * their own tests and the channel was dead in both directions: every envelope the shell sent was
 * refused by the gateway, and every reply the gateway sent was refused here. The serializer was
 * never wrong — `canonicalPayload` and Python's `canonical_payload` agree byte-for-byte on nested
 * objects, non-ASCII and key order. What was wrong was WHAT GOT HANDED TO IT.
 *
 * The lesson is written down here because this file is where it will be needed again: this module
 * is one half of a two-sided protocol, and a change to either half is a change to the contract.
 * `tests/unit/test_envelope_cross_language_parity.py` is the instrument that now fails when they
 * drift — it compares these functions' real output against the Python reference's on a fixed
 * hostile vector, so agreement is measured rather than assumed.
 *
 * Canonicalization matches Python's json.dumps(payload, sort_keys=True,
 * separators=(",",":"), ensure_ascii=False) for the payloads this channel actually carries:
 * objects with string keys and string/bool/int-or-null values (control ops). It is verified
 * byte-for-byte against the real Python gateway by the round-trip test (a Node-signed
 * envelope is accepted, and the gateway's reply verifies under the same key).
 *
 * Known, deliberately-unrelied-on edges where JS and Python could differ — and where the
 * difference is SAFE because it only makes a signature fail to verify (fail-closed): floats
 * (Python emits "1.0", JS "1") and astral-plane object keys (UTF-16 vs code-point sort).
 * This channel sends neither; if a future op needs them, add a cross-language numeric test
 * before relying on the tag rather than trusting these two encoders to agree.
 */
const crypto = require("node:crypto");

/** Deterministic canonical form the HMAC is taken over. Mirrors Python's sorted-key,
 *  compact, UTF-8 encoding exactly (recursively sorted object keys). */
function canonicalPayload(payload) {
  return Buffer.from(_canonicalize(payload), "utf8");
}

function _canonicalize(value) {
  if (value === null || typeof value !== "object") return JSON.stringify(value);
  if (Array.isArray(value)) return "[" + value.map(_canonicalize).join(",") + "]";
  const keys = Object.keys(value).sort();
  return "{" + keys.map((k) => JSON.stringify(k) + ":" + _canonicalize(value[k])).join(",") + "}";
}

/**
 * The bytes the HMAC is taken over: the WHOLE envelope except `integrity` itself. Mirrors Python's
 * `canonical_envelope`, including the decision to REUSE the one serializer rather than write a
 * second — two canonicalizations that can disagree is a signature bypass, and having two of them
 * per language would be four.
 */
function canonicalEnvelope(envelope) {
  const body = { ...(envelope || {}) };
  delete body.integrity;
  return canonicalPayload(body);
}

/** Mirrors Python's `compute_integrity(key, envelope)` — note the argument is the ENVELOPE. */
function computeIntegrity(key, envelope) {
  return crypto.createHmac("sha256", key).update(canonicalEnvelope(envelope)).digest("hex");
}

/** Constant-time verification — a wrong signature cannot advance (mirrors hmac.compare_digest). */
function verifyIntegrity(key, envelope) {
  const expected = computeIntegrity(key, envelope);
  const got = String((envelope && envelope.integrity) || "");
  if (expected.length !== got.length) return false;
  return crypto.timingSafeEqual(Buffer.from(expected, "utf8"), Buffer.from(got, "utf8"));
}

/**
 * How long a signed envelope may be acted on. Mirrors Python's `FRESHNESS_WINDOW_S`; the two are
 * pinned equal MECHANICALLY by the cross-language test, which reads both rather than trusting this
 * comment. Two integers do not need a configuration subsystem to stay in step — they need something
 * that fails when they stop being in step.
 */
const FRESHNESS_WINDOW_S = 300;

//: A timestamp with no zone designator has no defined instant, and JS would silently read it as
//: LOCAL time — so the window would mean whatever zone the reader happens to sit in. Python's
//: `fromisoformat` path refuses naive timestamps explicitly; this is that refusal, and it has to be
//: a syntax check because `Date.parse` will happily invent an answer. Accepts what Python 3.12
//: accepts: `Z`, `+HH`, `+HHMM`, `+HH:MM`.
const HAS_ZONE = /(?:[Zz]|[+-]\d{2}(?::?\d{2})?)$/;

/**
 * Is the SIGNED `ts` inside the freshness window? Fail closed on anything unparseable.
 *
 * Checked in BOTH directions, as Python does: an envelope from the future is as suspect as a stale
 * one, since a sender whose clock runs fast would otherwise mint envelopes with an arbitrarily long
 * life.
 */
function envelopeIsFresh(envelope, { now = null, windowS = FRESHNESS_WINDOW_S } = {}) {
  const raw = envelope && envelope.ts;
  if (typeof raw !== "string" || !HAS_ZONE.test(raw)) return false;
  const stamped = Date.parse(raw);
  if (Number.isNaN(stamped)) return false;
  const reference = now instanceof Date ? now.getTime() : Date.now();
  if (Number.isNaN(reference)) return false;
  return Math.abs(reference - stamped) <= windowS * 1000;
}

/** The envelope's fields, before it is signed. Split out so the signature is taken over the
 *  assembled object rather than computed inside the literal that builds it. */
function _unsignedEnvelope({ fromNode, to, msgType, payload, nodeCredential, scope,
                             taskId = null, payloadSchema = "control_event@1.0" }) {
  return {
    msg_id: crypto.randomUUID(),
    ts: new Date().toISOString(),
    schema: "envelope@1.0",
    from_node: fromNode,
    to,
    type: msgType,
    task_id: taskId,
    payload_schema: payloadSchema,
    payload,
    auth: { node_credential: nodeCredential, scope },
  };
}

/** Build a fully-signed envelope. integrity is computed here, never by callers. */
function makeEnvelope(spec) {
  // U458: assembled FIRST, then signed over itself. Computing the tag inline in the literal above
  // is what made the payload-only signature so easy to write and so hard to see — the object being
  // signed did not yet contain the fields the signature is supposed to cover. Python's
  // `make_envelope` was restructured the same way for the same reason.
  const envelope = _unsignedEnvelope(spec);
  envelope.integrity = computeIntegrity(spec.key, envelope);
  return envelope;
}

module.exports = {
  canonicalPayload, canonicalEnvelope, computeIntegrity, verifyIntegrity, makeEnvelope,
  envelopeIsFresh, FRESHNESS_WINDOW_S,
};

File: evidence/cp01/before/modules/sow/control_plane/ipc/envelope.py
"""Control-protocol envelope: schema validation + hmac-sha256 integrity (Plan 9.2, TB-2).

Envelopes are the ONLY control channel. Two independent checks make a message trustworthy,
and BOTH must pass before anything acts on it (fail-closed):

  1. structural: it validates against ``schemas/message.schema.json`` (``envelope@1.0``);
  2. integrity: ``integrity == hmac_sha256(shared_key, canonical(payload))`` — verified with
     a constant-time compare so a wrong signature cannot advance.

The message schema pins ``integrity`` to 64 hex chars but nothing in the tree computed the
HMAC until now; this module is that missing compute/verify pair. Canonicalization mirrors the
succession-snapshot convention (sorted keys, UTF-8) so signer and verifier agree byte-for-byte.
"""
from __future__ import annotations

import hashlib
import hmac
import json
import uuid
from datetime import datetime, timezone
from pathlib import Path
from typing import Any

import jsonschema

_SCHEMA_DIR = Path(__file__).resolve().parents[2] / "schemas"
_ENVELOPE_SCHEMA = json.loads((_SCHEMA_DIR / "message.schema.json").read_text(encoding="utf-8"))
_FMT = jsonschema.FormatChecker()  # enforce uuid/date-time formats (spec-audit F12 convention)
_VALIDATOR = jsonschema.Draft7Validator(_ENVELOPE_SCHEMA, format_checker=_FMT)


class EnvelopeError(Exception):
    """Envelope failed structural validation or integrity — fail closed, do not act on it."""


def canonical_payload(payload: dict[str, Any]) -> bytes:
    """Byte form the HMAC is taken over. Deterministic: sorted keys, compact, UTF-8."""
    return json.dumps(payload, sort_keys=True, separators=(",", ":"), ensure_ascii=False).encode("utf-8")


#: How long a signed envelope may be acted on. W-41: the `ts` is INSIDE the signature, so this is a
#: bound and not merely a claim the sender makes about itself. It is also what lets the replay cache
#: be bounded — an entry older than this can be dropped, because a replay of it now fails here.
FRESHNESS_WINDOW_S = 300


def canonical_envelope(envelope: dict[str, Any]) -> bytes:
    """Byte form the HMAC is taken over: the WHOLE envelope except `integrity` itself.

    W-41. This used to sign `payload` alone, so `from_node`, `type`, `task_id`, `msg_id` and `ts`
    were all rewritable in flight while the HMAC still verified — the routing and identity of a
    message were unauthenticated.

    It reuses `canonical_payload` rather than serialising separately. Two canonicalizations that
    disagree is a signature bypass, so there is exactly one serializer in this module and this
    function only decides WHAT is handed to it.
    """
    return canonical_payload({k: v for k, v in envelope.items() if k != "integrity"})


def compute_integrity(key: bytes, envelope: dict[str, Any]) -> str:
    return hmac.new(key, canonical_envelope(envelope), hashlib.sha256).hexdigest()


def verify_integrity(key: bytes, envelope: dict[str, Any]) -> bool:
    expected = compute_integrity(key, envelope)
    return hmac.compare_digest(expected, str(envelope.get("integrity", "")))


def envelope_is_fresh(envelope: dict[str, Any], *, now: datetime | None = None,
                      window_s: int = FRESHNESS_WINDOW_S) -> bool:
    """Is the SIGNED `ts` inside the freshness window? Fail closed on anything unparseable.

    Checked in BOTH directions: an envelope from the future is as suspect as a stale one, and a
    sender whose clock runs fast would otherwise mint envelopes with an arbitrarily long life.
    """
    raw = envelope.get("ts")
    if not isinstance(raw, str):
        return False
    try:
        stamped = datetime.fromisoformat(raw)
    except ValueError:
        return False
    if stamped.tzinfo is None:
        return False                      # a naive timestamp has no defined instant; refuse it
    reference = now or datetime.now(timezone.utc)
    return abs((reference - stamped).total_seconds()) <= window_s


def validate_structure(envelope: dict[str, Any]) -> None:
    """Raise EnvelopeError unless the object is a well-formed envelope@1.0."""
    errors = sorted(_VALIDATOR.iter_errors(envelope), key=lambda e: e.path)
    if errors:
        raise EnvelopeError(f"envelope failed schema: {errors[0].message}")


def make_envelope(
    *,
    from_node: str,
    to: list[str],
    msg_type: str,
    payload: dict[str, Any],
    node_credential: str,
    scope: str,
    key: bytes,
    task_id: str | None = None,
    payload_schema: str = "control_event@1.0",
) -> dict[str, Any]:
    """Build a fully-signed envelope. The integrity field is computed here, not by callers."""
    envelope = {
        "msg_id": str(uuid.uuid4()),
        "ts": datetime.now(timezone.utc).isoformat(),
        "schema": "envelope@1.0",
        "from_node": from_node,
        "to": to,
        "type": msg_type,
        "task_id": task_id,
        "payload_schema": payload_schema,
        "payload": payload,
        "auth": {"node_credential": node_credential, "scope": scope},
    }
    # W-41: signed AFTER assembly, over the whole envelope. Computing it inline above would have
    # signed a dict that did not yet contain the fields the signature is supposed to cover.
    envelope["integrity"] = compute_integrity(key, envelope)
    validate_structure(envelope)  # never emit a malformed envelope
    return envelope
